### PR TITLE
Store reasons for reboot transactions

### PR DESCRIPTION
**Pull request by j-mracek**
Requires: https://github.com/rpm-software-management/dnf/pull/1589

Based on: https://github.com/rpm-software-management/dnf-plugins-extras/pull/171

### CODE REVIEW

File: plugins/system_upgrade.py [hunk 1]
```diff
@@ -389,6 +392,30 @@ def _set_cachedir(self):
...
+                    forward.setdefault(pkg.repo.id, {}).setdefault(
+                        str(pkg), {})[tsi.action] = tsi.reason
+                except KeyError:
+                    raise RuntimeError("TransactionItem is has no RPM attached: %s" % tsi)
+            if tsi.action in dnf.transaction.BACKWARD_ACTIONS + \
+                    [libdnf.transaction.TransactionItemAction_REINSTALLED]:
+                try:
+                    reverse.setdefault(str(tsi.pkg), {})[tsi.action] = tsi.reason
+                except KeyError:
+                    raise RuntimeError("TransactionItem is has no RPM attached: %s" % tsi)
```
Comment (j-mracek) [inline plugins/system_upgrade.py]: delete try

File: plugins/system_upgrade.py [hunk 2]
```diff
@@ -389,6 +392,30 @@ def _set_cachedir(self):
...
+        reverse = {}
+        for tsi in self.cli.base.transaction:
+            if tsi.action in dnf.transaction.FORWARD_ACTIONS:
+                try:
+                    pkg = tsi.pkg
+                    forward.setdefault(pkg.repo.id, {}).setdefault(
+                        str(pkg), {})[tsi.action] = tsi.reason
+                except KeyError:
+                    raise RuntimeError("TransactionItem is has no RPM attached: %s" % tsi)
+            if tsi.action in dnf.transaction.BACKWARD_ACTIONS + \
```
Comment (j-mracek) [inline plugins/system_upgrade.py]: elif

File: plugins/system_upgrade.py [hunk 3]
```diff
@@ -626,14 +650,43 @@ def run_log(self):
...
+                try:
+                    pkg = tsi.pkg
+                    try:
+                        stored_reason = remove_packages[str(pkg)][str(tsi.action)]
+                        if stored_reason != tsi.reason:
+                            tsi.reason = stored_reason
+                    except KeyError:
+                        pass
+                except KeyError:
+                    raise RuntimeError("TransactionItem is has no RPM attached: %s" % tsi)
```
Comment (pkratoch) [inline plugins/system_upgrade.py]: This error is on more places than just the one and I suppose it's the same case.

File: plugins/system_upgrade.py [hunk 4]
```diff
@@ -626,14 +650,43 @@ def run_log(self):
...
+                    pkg = tsi.pkg
+                    try:
+                        stored_reason = install_packages[pkg.repo.id][str(pkg)][str(tsi.action)]
+                        if stored_reason != tsi.reason:
+                            tsi.reason = stored_reason
+                    except KeyError:
+                        pass
+                except KeyError:
+                    raise RuntimeError("TransactionItem is has no RPM attached: %s" % tsi)
+            if tsi.action in dnf.transaction.BACKWARD_ACTIONS + \
```
Comment (pkratoch) [inline plugins/system_upgrade.py]: Also "elif"

**Pull request merged**